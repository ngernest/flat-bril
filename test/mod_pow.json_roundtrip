{
  "functions": [
    {
      "args": [
        {
          "name": "base",
          "type": "int"
        },
        {
          "name": "exp",
          "type": "int"
        },
        {
          "name": "mod",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "base",
            "exp",
            "mod"
          ],
          "dest": "out",
          "funcs": [
            "mod_pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "out"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "e",
          "type": "int"
        },
        {
          "name": "m",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "m"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "dest": "v5",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v5"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "base",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "base"
          ],
          "dest": "base",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "e"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9"
          ],
          "dest": "exp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.8"
        },
        {
          "args": [
            "exp"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v12"
          ],
          "funcs": [],
          "labels": [
            "for.body.8",
            "for.end.8"
          ],
          "op": "br"
        },
        {
          "label": "for.body.8"
        },
        {
          "args": [
            "exp"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "bit",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "bit"
          ],
          "dest": "bit",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "bit"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v17",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v18"
          ],
          "funcs": [],
          "labels": [
            "then.15",
            "else.15"
          ],
          "op": "br"
        },
        {
          "label": "then.15"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "base"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v19",
            "v20"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v21"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v23",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v22",
            "v23"
          ],
          "dest": "v24",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v24"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.15"
          ],
          "op": "jmp"
        },
        {
          "label": "else.15"
        },
        {
          "label": "endif.15"
        },
        {
          "args": [
            "base"
          ],
          "dest": "v25",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "base"
          ],
          "dest": "v26",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v25",
            "v26"
          ],
          "dest": "v27",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v27"
          ],
          "dest": "base",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "base"
          ],
          "dest": "v28",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v29",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v28",
            "v29"
          ],
          "dest": "v30",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v30"
          ],
          "dest": "base",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "exp"
          ],
          "dest": "v31",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v32",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v31",
            "v32"
          ],
          "dest": "v33",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v33"
          ],
          "dest": "exp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.8"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.8"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v34",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v34"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod_pow",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "m",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
