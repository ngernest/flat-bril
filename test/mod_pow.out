@main
	op: call 		dest: out		type: int		args: ["base", "exp", "mod"]		funcs: ["mod_pow"]
	op: print		args: ["out"]	
@mod_pow
	op: id   		dest: v1		type: int		args: ["m"]	
	op: const		dest: v2		type: int		value: 1	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: br   		args: ["v3"]		labels: ["then.0", "else.0"]
.then.0
	op: const		dest: v4		type: int		value: 0	
	op: ret  		args: ["v4"]	
.else.0
	op: const		dest: v5		type: int		value: 1	
	op: id   		dest: result		type: int		args: ["v5"]	
	op: id   		dest: v6		type: int		args: ["b"]	
	op: id   		dest: v7		type: int		args: ["m"]	
	op: call 		dest: base		type: int		args: ["v6", "v7"]		funcs: ["mod"]
	op: id   		dest: base		type: int		args: ["base"]	
	op: id   		dest: v9		type: int		args: ["e"]	
	op: id   		dest: exp		type: int		args: ["v9"]	
.for.cond.8
	op: id   		dest: v10		type: int		args: ["exp"]	
	op: const		dest: v11		type: int		value: 0	
	op: gt   		dest: v12		type: bool		args: ["v10", "v11"]	
	op: br   		args: ["v12"]		labels: ["for.body.8", "for.end.8"]
.for.body.8
	op: id   		dest: v13		type: int		args: ["exp"]	
	op: const		dest: v14		type: int		value: 2	
	op: call 		dest: bit		type: int		args: ["v13", "v14"]		funcs: ["mod"]
	op: id   		dest: bit		type: int		args: ["bit"]	
	op: id   		dest: v16		type: int		args: ["bit"]	
	op: const		dest: v17		type: int		value: 1	
	op: eq   		dest: v18		type: bool		args: ["v16", "v17"]	
	op: br   		args: ["v18"]		labels: ["then.15", "else.15"]
.then.15
	op: id   		dest: v19		type: int		args: ["result"]	
	op: id   		dest: v20		type: int		args: ["base"]	
	op: mul  		dest: v21		type: int		args: ["v19", "v20"]	
	op: id   		dest: result		type: int		args: ["v21"]	
	op: id   		dest: v22		type: int		args: ["result"]	
	op: id   		dest: v23		type: int		args: ["m"]	
	op: call 		dest: v24		type: int		args: ["v22", "v23"]		funcs: ["mod"]
	op: id   		dest: result		type: int		args: ["v24"]	
	op: jmp  		labels: ["endif.15"]
.else.15
.endif.15
	op: id   		dest: v25		type: int		args: ["base"]	
	op: id   		dest: v26		type: int		args: ["base"]	
	op: mul  		dest: v27		type: int		args: ["v25", "v26"]	
	op: id   		dest: base		type: int		args: ["v27"]	
	op: id   		dest: v28		type: int		args: ["base"]	
	op: id   		dest: v29		type: int		args: ["m"]	
	op: call 		dest: v30		type: int		args: ["v28", "v29"]		funcs: ["mod"]
	op: id   		dest: base		type: int		args: ["v30"]	
	op: id   		dest: v31		type: int		args: ["exp"]	
	op: const		dest: v32		type: int		value: 2	
	op: div  		dest: v33		type: int		args: ["v31", "v32"]	
	op: id   		dest: exp		type: int		args: ["v33"]	
	op: jmp  		labels: ["for.cond.8"]
.for.end.8
	op: id   		dest: v34		type: int		args: ["result"]	
	op: ret  		args: ["v34"]	
@mod
	op: id   		dest: v0		type: int		args: ["n"]	
	op: id   		dest: v1		type: int		args: ["n"]	
	op: id   		dest: v2		type: int		args: ["m"]	
	op: div  		dest: v3		type: int		args: ["v1", "v2"]	
	op: id   		dest: v4		type: int		args: ["m"]	
	op: mul  		dest: v5		type: int		args: ["v3", "v4"]	
	op: sub  		dest: v6		type: int		args: ["v0", "v5"]	
	op: ret  		args: ["v6"]	
